sdk/tests/conformance/extensions/khr-parallel-shader-compile.html

```diff
@@ -58,36 +95,131 @@
-
-    debug("Checking that status is true when context is lost.");
-    if (loseContext) {
-        loseContext.loseContext();
-        status = gl.getShaderParameter(vs, COMPLETION_STATUS_KHR);
-        if (status !== true) testFailed("status should be true when context is lost");
+(async () => {
+
+    if (!ext) {
+        testPassed("No KHR_parallel_shader_compile support -- this is legal");
+    } else {
+        testPassed("Successfully enabled KHR_parallel_shader_compile extension");
+
+        shouldBe("ext.COMPLETION_STATUS_KHR", "0x91B1");
+
+        debug("Checking that status is a boolean.");
+        gl.shaderSource(vs, vertexSource());
+        gl.compileShader(vs);
+        let status = gl.getShaderParameter(vs, COMPLETION_STATUS_KHR);
+        if (status !== true && status !== false) testFailed("status should be a boolean");
+
+        gl.linkProgram(program);
         status = gl.getProgramParameter(program, COMPLETION_STATUS_KHR);
-        if (status !== true) testFailed("status should be true when context is lost");
+        if (status !== true && status !== false) testFailed("status should be a boolean");
+
+        debug("Constructing shader that takes >500ms to compile");
+        let measuredCompileDuration = 0;
+        extraCode = '\n    if (true) { result += vec4(0.0000001); }';
+        for (let i = 0; measuredCompileDuration < 500; i++) {
+            extraCode += extraCode;
+            extraCode += extraCode;
+            if (i < 4) continue;
+            gl.shaderSource(vs, vertexSource(extraCode));
+            gl.shaderSource(fs, fragmentSource(extraCode));
+            gl.compileShader(vs);
+            gl.compileShader(fs);
+            gl.linkProgram(program);
+            const start = performance.now();
+            if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
+                testFailed(`Shaders failed to compile.
+                            program: ${gl.getProgramInfoLog(program)}
+                            vs: ${gl.getShaderInfoLog(vs)}
+                            fs: ${gl.getShaderInfoLog(fs)}`);
+                break;
+            }
+            measuredCompileDuration = performance.now() - start;
+        }
+
+        debug('');
+        gl.shaderSource(vs, vertexSource(extraCode));
+        gl.shaderSource(fs, fragmentSource(extraCode));
+        gl.compileShader(vs);
+        gl.compileShader(fs);
+        gl.linkProgram(program);
+
+        let start = performance.now();
+        gl.getShaderParameter(fs, COMPLETION_STATUS_KHR);
+        gl.getShaderParameter(vs, COMPLETION_STATUS_KHR);
+        let duration = performance.now() - start;
+        if (performance.now() - start > 100)
+            testFailed(`Querying shader status should not wait for compilation. Took ${duration} ms`);
+
+        let frames = 0;
+        while (!gl.getProgramParameter(program, COMPLETION_STATUS_KHR)
+            && performance.now() - start < measuredCompileDuration * 4) {
```

Comment:
Use a separate for the factor?

Especially concerned about performance on slower machines, such as those in Chromium's testing fleet. I think this will end up being very flaky.

Reply:
I agree that flakiness is a danger, but I think with some tuning on the thresholds it can be made quite reliable. It shouldn't matter how slow or fast the machine is in absolute terms, because we're comparing it to itself. This would only be flaky if the shader takes more than 4 times longer to compile the second time vs the first time.

Reply:
Other tests - conformance/rendering/texture-switch-performance.html specifically - have attempted to compare against baseline performance but were not successful, and were ultimately too flaky to be turned on on Chromium's test bots. See the blanket suppression for that test in Chromium's webgl_conformance_expectations.txt.

We can certainly try this on the bots, though I'd suggest factoring out the factor into a separate constant.